crates/ruma-identifiers/src/transaction_id.rs

```diff
@@ -0,0 +1,36 @@
+/// A Matrix transaction ID.
+///
+/// Transaction IDs in Matrix are opaque strings. This type is provided simply for its semantic
+/// value.
+///
+/// You can create one from a string (using `.into()`) but the recommended way is to use
+/// `TransactionId::new()` to generate a random one. If that function is not available for you, you
+/// need to activate this crate's `rand` Cargo feature.
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TransactionId(str);
+
+impl TransactionId {
+    /// Creates a random transaction ID.
+    ///
+    /// This will currently be a UUID without hyphens, but no guarantees are made about the
+    /// structure of transaction IDs generated from this function.
+    #[cfg(feature = "rand")]
+    pub fn new() -> Box<Self> {
+        let id = uuid::Uuid::new_v4();
+
+        let mut buf: Box<[u8]> = Box::new([0; 32]);
+        id.to_simple().encode_lower(&mut buf);
+
+        // SAFETY: str has the same representation as [u8], as guaranteed by
+        // `str::into_boxed_bytes`¹ saying "without copying or allocating".
+        //
+        // ¹ https://doc.rust-lang.org/std/primitive.str.html#method.into_boxed_bytes
```

Comment:
In general i believe this can be done better by [String::into_boxed_str](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.into_boxed_str), as we're already allocating something.

All this unsafe stuff, plus allocating a box, can be replaced with;

```rs
let id = uuid::Uuid::new_v4();

id.to_simple().to_string().into_boxed_str()

Self::from_owned(buf_str)
```

Reply:
Yeah, should be good enough I guess.